### PR TITLE
TOOLS-2459 Allow --uri to be used with other connection options

### PR DESCRIPTION
**Pull request by tfogo**
This PR refactors the `ToolOptions.ParseArgs()` code path to allow `--uri` to be used with other connection options. Connection options can be set either in the URI or in other CLI options. The tools will then combine these options to configure the client connection. If options are specified in more than one place an error will be thrown if they conflict in some way. No error is thrown if they are the same. 

## The zero value problem

The current implementation doesn't throw a conflict in all the cases that it should. This is because the way we are using the go-flags package means we can't tell for sure if an option has been set or not. The closest approximation is by checking if the option is different from its default value. But this causes an issue if a user intentionally set the default value of an option.

For example the following combination of arguments would not throw an error even though clearly there has been a mistake:

```
--dialTimeout 3 --uri="mongodb://foo/?connectTimeoutMS=5000"
--compressors none --uri="mongodb://foo/?compressors=snappy"
--ssl false --uri="mongodb://foo/?tls=true"
```

There are several possible solutions to this:
- Use pointers for all option values.
- Use go-flags' Command objects so we have access to their Option type. This type can tell us whether or not it has been set.
- Use a custom option type.
- Switch to a different CLI framework such as spf13/cobra or urfave/cli (probably a bit drastic).

I'm going to open a separate ticket to find a solution to these zero value problems. 

## Some changes from the Design Doc

- Originally this should have thrown a conflict:
    
    ```
    --port 30000 --uri="mongodb://foo"
    ```

    This is because we assume that if a port is not set on the connection string, the default is 27017. When I was implementing this, I became convinced that this would be extremely frustrating for users. I think the assumption is that if you intentionally define an option, that should override a default. It's definitely confusing to tell people that their port of 30000 conflicts with the default port of 27017 when 27017 isn't even in the connection string. By far the most used options with `--uri` will be `--username`, `--password`, `--db`, and `--port`, so I think it's important that we ensure the usage of `--port` is not jarring.
- The design doc only listed a subset of the CLI options that could potentially conflict with URIs, so I tried to deal with all the options. I had to ignore some, which are detailed below. 

## Notes on options

I definitely have a deeper appreciation for the need to rationalize tools options. There are options which are misnamed, options which do nothing, options which do more than one thing, and even multiple options that do the same thing. for reference I've made a table of all connection-related options. 

| CLI Option  | Default Value  | Corresponding URI Parameter  | Notes   | 
|---|---|---|---|
| `--db`  | `""` | In URI  |   | 
| `--host`  | `""`  | In URI, also contains port and `replSet`  |   | 
| `--port`  | `""`  |  In URI |   | 
| `--dialTimeout`  |  `3` | `connectTimeoutMS`  |   | 
| `--socketTimeout`  |  `0` |  `socketTimeoutMS` |   | 
| `--TCPKeepAliveSeconds`  |  `30` | Nothing documented  | **Conflict checking not implemented.** As far as I can tell this is unused.  | 
| `--serverSelectionTimeout`  |  `0` | `serverSelectionTimeoutMS` |   | 
| `--compressors`  |  `"none"` |  `compressors` |   | 
| `--ssl`  |  `false` | `ssl`/`tls` (also set by default with SRV scheme)  |   | 
| `--sslCAFile`  |  `""` | `sslCertificateAuthorityFile`/`tlsCAFile`  |   | 
| `--sslPEMKeyFile`  |  `""` | `sslClientCertificateKeyFile`/`tlsCertificateKeyFile`  |   | 
| `--sslPEMKeyPassword`  |  `""` | `sslClientCertificateKeyPassword`/`tlsCertificateKeyFilePassword`  |   | 
| `--sslCRLFile`  |  `""` |  Nothing documented | **Conflict checking not implemented.**  | 
| `--sslAllowInvalidCertificates`  |  `false` | `sslInsecure`/`tlsInsecure`  |   | 
| `--sslAllowInvalidHostnames`  |  `false` | `sslInsecure`/`tlsInsecure`  |   | 
| `--sslFIPSMode`  |  `false` | Nothing documented | **Conflict checking not implemented.**  | 
| `--username`  |  `""` | In URI |   | 
| `--password`  |  `""` | In URI |   | 
| `--authenticationDatabase`  |  `""` | `authSource`  |   | 
| `--authenticationMechanism`  |  `""` | `authMechanism`  |   | 
| `--gssapiServiceName`  |  `""` | `authMechanismProperties=SERVICE_NAME:`  |   | 
| `--gssapiHostName`  |  `""` | Nothing documented | **Conflict checking not implemented.** As far as I can tell this is unused.  | 
| `--writeConcern`  |  `"majority"` |  `w`/`journal`/`wTimeoutMS` | **Conflict checking not implemented.** The `options.WriteConcern` struct isn't used anywhere. mongofiles has its own implementation.  | 
| `--readPreference`  |  `""` |  `readPreference` | **Conflict checking not implemented.** Only for mongoexport, mongodump, and mongofiles  | 

## Implementation

Ideally I'd like to have created a connString object from toolOptions then merge it with a connString object from a URI. However, the zero value problem really gets in the way. I think there will be a much more streamlined way of implementing all this once we have some way of telling when options have actually been set. 

For the TOOLS_TESTING_SRV test type, how do we set an ATLAS_URI for the evergreen tests? I'd like to avoid putting a URI in the evergreen.yml.

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -26,7 +26,8 @@
 
 [[constraint]]
   name = "go.mongodb.org/mongo-driver"
-  version = "^v1.1.2"
+  branch = "GODRIVER-1490"
+  source = "github.com/tfogo/mongo-go-driver"
```
Comment (pmeredit) [inline Gopkg.toml]: Why is this necessary? Just curious.

File: options/options.go [hunk 2]
```diff
@@ -44,7 +44,10 @@ var (
 )
 
 const IncompatibleArgsErrorFormat = "illegal argument combination: cannot specify %s and --uri"
-const ConflictingArgsErrorFormat = "illegal argument combination: %s conflicts with --uri"
+
+func ConflictingArgsErrorFormat(optionName, uriValue, cliValue, cliOptionName string) error {
+	return fmt.Errorf("Invalid Options: Cannot specify different %s in connection URI and command-line option (\"%s\" was specified in the URI and \"%s\" was specified in the %s option)", optionName, uriValue, cliValue, cliOptionName)
+}
```
Comment (pmeredit) [inline options/options.go]: It might be beneficial to keep a mapping from cliOptionName to optionName so that you don't need to pass both arguments, but I'm fine with this.

File: options/options.go [hunk 3]
```diff
@@ -460,24 +463,33 @@ func (opts *ToolOptions) ParseArgs(args []string) ([]string, error) {
...
-func (opts *ToolOptions) NormalizeHostPortURI() error {
+// NormalizeOptionsAndURI syncs the connection string an toolOptions objects.
+// It returns an error if there is any conflict between options and the connection string.
+// If a value is set on the options, but not the connection string, that value is added to the
+// connection string. If a value is set on the connection string, but not the options,
+// that value is added to the options.
+func (opts *ToolOptions) NormalizeOptionsAndURI() error {
 	if opts.URI != nil && opts.URI.ConnectionString != "" {
-		cs, err := connstring.Parse(opts.URI.ConnectionString)
+		cs, err := connstring.ParseWithoutValidating(opts.URI.ConnectionString)
```
Comment (pmeredit) [inline options/options.go]: This actually would have helped to give better error messages in one of the Mirror tickets I did, nice!

File: options/options.go [hunk 4]
```diff
@@ -503,80 +515,282 @@ func (opts *ToolOptions) handleUnknownOption(option string, arg flags.SplitArgum
...
-			return fmt.Errorf(IncompatibleArgsErrorFormat, "--socketTimeout")
-		// TODO: TOOLS-2348 will need to address this new argument
-		case opts.Connection.Compressors != "" && opts.Connection.Compressors != "none":
-			return fmt.Errorf(IncompatibleArgsErrorFormat, "--compressors")
-		}
-		opts.Connection.Timeout = int(cs.ConnectTimeout / time.Millisecond)
-		opts.Connection.SocketTimeout = int(cs.SocketTimeout / time.Millisecond)
+		// Port can be set in --port, --host, or URI
+		// Each host/port pair in the options must match the URI host/port pairs
+		if opts.Port != "" {
```
Comment (pmeredit) [inline options/options.go]: Since this is done, it's done :) But this would have probably been a good place to use codegen from a json/yaml file. It's right on the boundary for size.

File: options/options_test.go [hunk 5]
```diff
@@ -420,6 +446,183 @@ func TestParseAndSetOptions(t *testing.T) {
...
+			{"--host foo", "mongodb://bar", ShouldFail},
+			{"--port 27018", "mongodb://foo", ShouldSucceed},
+			{"--port 27018", "mongodb://foo:27019", ShouldFail},
+			{"--port 27018", "mongodb://foo:27018", ShouldSucceed},
+			{"--host foo:27019 --port 27018", "mongodb://foo", ShouldFail},
+			{"--host foo:27018 --port 27018", "mongodb://foo:27018", ShouldSucceed},
+			{"--host foo:27019 --port 27018", "mongodb://foo:27018", ShouldFail},
+
+			{"--host foo,bar,baz", "mongodb://foo,bar,baz", ShouldSucceed},
+			{"--host foo,bar,baz", "mongodb://baz,bar,foo", ShouldSucceed},
```
Comment (pmeredit) [inline options/options_test.go]: nice!

File: options/options_test.go [hunk 6]
```diff
@@ -420,6 +446,183 @@ func TestParseAndSetOptions(t *testing.T) {
...
+			{"--host repl/foo,bar,baz", "mongodb://foo,bar,baz/?replicaSet=repl", ShouldSucceed},
+			{"--host repl/foo,bar,baz", "mongodb://foo,bar,baz/?replicaSet=quux", ShouldFail},
+
+			// Compressors
+			{"--compressors snappy", "mongodb://foo/?compressors=snappy", ShouldSucceed},
+			{"", "mongodb://foo/?compressors=snappy", ShouldSucceed},
+			{"--compressors snappy", "mongodb://foo/", ShouldSucceed},
+			{"--compressors snappy", "mongodb://foo/?compressors=zlib", ShouldFail},
+			// {"--compressors none", "mongodb://foo/?compressors=snappy", ShouldFail}, // Note: zero value problem
+			{"--compressors snappy", "mongodb://foo/?compressors=none", ShouldFail},
```
Comment (pmeredit) [inline options/options_test.go]: I think it'd be worth doing the `Option` route... in another ticket.

File: options/options.go [hunk 7]
```diff
@@ -503,80 +515,282 @@ func (opts *ToolOptions) handleUnknownOption(option string, arg flags.SplitArgum
...
+	if opts.ReplicaSetName != "" && cs.ReplicaSet != "" {
+		if opts.ReplicaSetName != cs.ReplicaSet {
+			return ConflictingArgsErrorFormat("replica set name", cs.ReplicaSet, opts.Host, "--host")
+		}
+
+	}
+	if opts.ReplicaSetName != "" && cs.ReplicaSet == "" {
+		cs.ReplicaSet = opts.ReplicaSetName
+	}
+	if opts.ReplicaSetName == "" && cs.ReplicaSet != "" {
```
Comment (pmeredit) [inline options/options.go]: This could be an `else if`, but I'm not too worried about it. (Ditto the other places this pattern is used).

Comment (pmeredit) [reply options/options.go]: Alternatively, you can also use switches, which have the same benefit while looking slightly less cluttered.

File: Gopkg.toml (hunk 1 above)
Comment (rychipman) [reply Gopkg.toml]: I think it's because the go driver PR hasn't been merged yet

File: options/options.go [hunk 8]
```diff
@@ -460,24 +463,33 @@ func (opts *ToolOptions) ParseArgs(args []string) ([]string, error) {
...
+	err = opts.NormalizeOptionsAndURI()
 	if err != nil {
 		return []string{}, err
 	}
 
 	return args, err
 }
 
-func (opts *ToolOptions) NormalizeHostPortURI() error {
+// NormalizeOptionsAndURI syncs the connection string an toolOptions objects.
```
Comment (rychipman) [inline options/options.go]:
```suggestion
// NormalizeOptionsAndURI syncs the connection string and toolOptions objects.
```

**Comment by tfogo** [comment]
@rychipman Looks like those failures are related to the bar writer and have been around for a while. I don't think they're related.

**Comment by rychipman** [comment]
Ah yes, you're totally right. In that case, LGTM!

File: options/options.go (hunk 4 above)
Comment (tfogo) [reply options/options.go]: Good point. I think we can look into that in the ticket for fixing the zero-value problem.

File: options/options.go (hunk 7 above)
Comment (tfogo) [reply options/options.go]: Yeah it's very cluttered currently. I'm hoping we can get rid of most of these once we fix the zero value problem.

**Comment by tfogo** [comment]
Okay, I confirmed David is fine with the design changes. I'm going to wait until the Go driver PR is in before merging, since I don't want to point to an unofficial repo.

**Comment by boutell** [comment]
Does this mean that `--db=foo` can be used to override a placeholder db name present in the connection URI (example: an Atlas URI will always have one)? That was the goal of #TOOLS-2378 which has been closed with reference to this PR, so I thought I'd check.

Thanks for the hard work!

**Comment by tfogo** [comment]
Hi @boutell, I'm afraid this won't let you have a different database in the connection string and the `--db` option. Previously, the tools wouldn't let you set the `--db` option if you're using `--uri`, even if there was no database in the URI. This change means you can have a URI with no database and set one via the `--db` option.

Any option that you set on the command line that conflicts with the URI will cause a failure. We spent a lot of time thinking about this feature and we decided that this was the safest way of doing it. This is particularly important for the database. Any ambiguity in which database is being acted upon could be very dangerous. Imagine a command like: 

```
mongorestore --drop --uri "mongodb://host/foo" --db bar`
```

If a user doesn't understand the order of overwriting, or doesn't notice that the databases are different, this has the potential to be disastrous.

So this PR really only resolves part of TOOLS-2378, sorry about that. It does mean that you can now strip the db off the connection string and specify any db you like in the `--db` option. I'm afraid overwriting the connection string db with the `--db` option is a "won't fix" given the safety concerns.

I hope this PR will at least ease some pain points. Thanks for your question!

**Comment by boutell** [comment]
I see your point, and the clarity is appreciated. I think, then, that Atlas should start offering its URLs without a "/test" database name stuffed in there. But that's an Atlas issue, so I'll forward it to them. Thanks.
